src/registry.rs

```diff
@@ -0,0 +1,249 @@
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! ObjectStoreRegistry holds object stores at runtime with a URL for each store.
+//! The registry serves as a cache for object stores to avoid repeated creation.
+//! It also lets you convert an [`ObjectStore`] back to its registered URL via `get_url`:
+//!
+//! ```rust
+//! use std::sync::Arc;
+//! use url::Url;
+//! use object_store::ObjectStore;
+//! use object_store::memory::InMemory;
+//! use object_store::registry::{DefaultObjectStoreRegistry, ObjectStoreRegistry};
+//!
+//! let store = Arc::new(InMemory::new()) as Arc<dyn ObjectStore>;
+//! let registry = DefaultObjectStoreRegistry::new();
+//! let url = Url::parse("inmemory://").unwrap();
+//! registry.register_store(&url, Arc::clone(&store));
+//! let found_store = registry.get_store(&url).unwrap();
+//! assert!(Arc::ptr_eq(&found_store, &store));
+//! let found_url = registry.get_url(Arc::clone(&store)).unwrap();
+//! assert_eq!(found_url, url);
+//! ```
+use crate::ObjectStore;
+use std::collections::HashMap;
+use std::sync::{Arc, RwLock};
+use url::Url;
+
+/// [`ObjectStoreRegistry`] maps a URL prefix to an [`ObjectStore`] instance. The definition of
+/// a URL prefix depends on the [`ObjectStoreRegistry`] implementation. See implementation docs
+/// for more details.
+pub trait ObjectStoreRegistry: Send + Sync + std::fmt::Debug + 'static {
+    /// Register a new store for any URL that begins with `prefix`
+    ///
+    /// If a store with the same prefix existed before, it is replaced and returned
+    fn register_store(
+        &self,
+        prefix: &Url,
+        store: Arc<dyn ObjectStore>,
+    ) -> Option<Arc<dyn ObjectStore>>;
+
+    /// Get a store for the provided URL. The definition of a "suitable store" depends on
+    /// the [`ObjectStoreRegistry`] implementation. See implementation docs for more details.
+    ///
+    /// If no [`ObjectStore`] is found for the `url`, an [`ObjectStore`] may be lazily be
+    /// created and registered. The logic for doing so is left to each [`ObjectStoreRegistry`]
+    /// implementation.
+    fn get_store(&self, url: &Url) -> Option<Arc<dyn ObjectStore>>;
+
+    /// Given one of the `Arc<dyn ObjectStore>`s you registered, return its URL.
+    fn get_url(&self, store: Arc<dyn ObjectStore>) -> Option<Url>;
+
+    /// List all registered store prefixes
+    fn get_store_prefixes(&self) -> Vec<Url>;
```

Comment:
one final pedantic question -- should we make it `Vec<&Url>`? Owned is probably fine though

Reply:
Eh, I kinda prefer keeping it Url just so we don't have to deal with lifetimes in the traits. If you feel strongly, I don't mind updating it though. :)

Reply:
I think it is fine